### PR TITLE
Reland Use scatter_reduce to support masked reductions on COO tensors (sum, prod, amin, amax)

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75454
* #75651

### CODE REVIEW

File: torch/_masked/__init__.py [hunk 1]
```diff
@@ -686,44 +775,28 @@ def sum(input: Tensor,
         mask: Optional[Tensor] = None) -> Tensor:
     # __doc__ is generated by _apply_docstring_templates decorator
     if dtype is None:
-        dtype = input.dtype
```
Comment (mikaylagawarecki) [inline torch/_masked/__init__.py:689]: Should we be promoting int to int64 for masked_sum? Noticed that `sum` seems to do this but we have `promotes_int_to_int64=False` in the OpInfo for `_masked.sum`

Comment (cpuhrsch) [reply torch/_masked/__init__.py:689]: Yes, we should mirror the behavior of torch.sum. @pearu - are we missing something here?

File: test/test_sparse.py [hunk 2]
```diff
@@ -3575,6 +3576,12 @@ def test_future_empty_dim(self, device, dtype, op):
 
             t = sample_input.input
             mask = sample_input_kwargs.get('mask')
+            if mask is None and op_name in {'prod', 'amax', 'amin'}:
+                # FIXME: for now reductions with non-zero reduction identity and
+                # unspecified mask are not supported for sparse COO
+                # tensors, see torch._masked.prod implementation
```
Comment (cpuhrsch) [inline test/test_sparse.py:3582]: But doesn't this PR introduce that exact support?

File: torch/_masked/__init__.py [hunk 3]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
 
 
+def _sparse_coo_scatter_reduction_helper(op,
+                                         mask_input: Tensor,
+                                         dims: Tuple[int, ...],
+                                         keepdim: bool,
+                                         dtype: Optional[DType] = None) -> Tensor:
+    reduce = op.__name__
+    assert reduce in {'sum', 'prod', 'mean', 'amax', 'amin'}, \
+        "op must be one of torch.sum, torch.prod, torch.mean, torch.amax or torch.amin"
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:522]: Also include the value that was actually received. In general I consider this a ValueError.

```
valid_reductions = ['sum', 'prod', ..., 'amin']
if op not in valid_reductions:
    throw ValueError(f"op must be one of {' '.join(valid_reductions}}, but got {op.__name__} instead")
```

File: torch/_masked/__init__.py [hunk 4]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
+    reduced_sparse_dims = []
+    retained_sparse_dims = []
+    reduced_dense_dims = []
+
+    # promote dtype if specified
+    if values.dtype != output_dtype:
+        values = values.to(output_dtype)
+
+    if keepdim:
+        output_shape = tuple((1 if i in dims else mask_input.shape[i]) for i in range(mask_input.ndim))
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:537]: This might be easier if you use enumerate
```
if keepdim:
    output_shape = tuple(1 if i in dims else si for (i, si) in enumerate(mask_input.shape))
else:
    output_shape = tuple(si for (i, si) in enumerate(mask_input.shape) if i not in dims)
```

File: torch/_masked/__init__.py [hunk 5]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
+            reduced_dense_dims.append(d + 1 - num_sparse_dims)
+
+    # Reduce dense dimensions
+    if len(reduced_dense_dims) > 0:
+        if reduce == "sum":
+            new_values = values
+            new_values = op(new_values, dim=reduced_dense_dims, keepdim=bool(keepdim))
+        else:
+            # FIXME: Implement reductions for dense dimensions for ops with non-zero reduction identities
+            raise NotImplementedError(f'{reduce}() does not support reducing dense dims of hybrid sparse coo tensors')
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:557]: Instead of raising a NotImplementedError it can be more convenient to return [`NotImplemented`](https://docs.python.org/3/library/constants.html?highlight=notimplemented#NotImplemented), which works better with our implementation of `__torch_dispatch__` while we're closing this gap.

File: torch/_masked/__init__.py [hunk 6]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
+    else:
+        new_values = values.clone()
+
+    # Reduce sparse dimensions
+    if len(reduced_sparse_dims) == num_sparse_dims:
+        if reduce in {'amax', 'amin'} and new_values.size(0) == 0:
+            # IndexError: amax(): Expected reduction dim 0 to have non-zero size.
+            # sum()/prod() return the reduction identity when dim has size 0 but amax()/amin() do not
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:565]: I wonder if there is an existing issue about this. If not, maybe you can open one and reference it here? Seems like a potentially useful feature.

File: torch/_masked/__init__.py [hunk 7]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
+    if (new_indices.numel() > 0):
+        # lexsort indices and get index tensor for scatter reduction
+        new_indices, inverse_indices = torch.unique(new_indices, return_inverse=True, dim=1)
+        out_shape = list(new_values.shape)
+        out_shape[0] = new_indices.shape[1]
+        for _ in range(new_values.ndim - 1):
+            inverse_indices = inverse_indices.unsqueeze(-1)
+        scatter_indices = inverse_indices.expand(new_values.shape)
+        # FIXME: temporary workaround for issue with bfloat16 remove when acctype is implemented for scatter_reduce
+        if output_dtype in {torch.bfloat16}:
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:592]: Is this needed for float16 as well?

File: torch/_masked/__init__.py [hunk 8]
```diff
@@ -686,44 +775,28 @@ def sum(input: Tensor,
         mask: Optional[Tensor] = None) -> Tensor:
     # __doc__ is generated by _apply_docstring_templates decorator
     if dtype is None:
-        dtype = input.dtype
+        # promote integer types to int64 when output dtype is not specified
+        if input.layout == torch.sparse_csr:
+            if input.dtype in {torch.uint8, torch.bool, torch.int8, torch.int16, torch.int32}:
+                # csr.to(dtype=torch.int64) is not implemented, so
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:781]: Can you open an issue for this missing feature if it doesn't exist yet? Ideally we just fix this before landing this.

Comment (cpuhrsch) [reply torch/_masked/__init__.py:781]: Bonus points for simply sending a PR that implements it. It seems like it should be simply about forwarding the call to `to` to the values of the CSR Tensor.

File: torch/_masked/__init__.py [hunk 9]
```diff
@@ -752,11 +825,15 @@ def prod(input: Tensor,
     if input.layout == torch.strided:
         # Workaround https://github.com/pytorch/pytorch/issues/56586
         result = mask_input
+        result = result.to(dtype=dtype)
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:828]: Why this change?

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -1107,6 +1108,12 @@ def sample_inputs_sparse_coo_masked_reduction(op_info, device, dtype, requires_g
                 inputs.append(SampleInput(sample_input.input.to_sparse(),
                                           args=sample_input.args, kwargs=sample_input_kwargs))
             else:
+                if op_name in {'prod', 'amax', 'amin'}:
+                    # FIXME: for now reductions with non-zero reduction identity and
+                    # unspecified mask are not supported for sparse COO
+                    # tensors, see torch._masked.prod implementation
+                    # for details.
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_methods_invocations.py:1115]: Isn't this only true when there's dense dimensions?

File: torch/_masked/__init__.py (hunk 8 above)
Comment (mikaylagawarecki) [reply torch/_masked/__init__.py:781]: This is actually copied from [the comment Pearu added in prod](https://github.com/pytorch/pytorch/blob/master/torch/_masked/__init__.py#L737-L749). Sure I can implement `to` for CSR

File: torch/_masked/__init__.py (hunk 9 above)
Comment (mikaylagawarecki) [reply torch/_masked/__init__.py:828]: I think the dtype parameter is supposed to >>> input tensor is casted to dtype before the operation is performed

over here I believe the casting was being done after so I changed it

File: test/test_sparse.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/test_sparse.py:3582]: When the mask is *unspecified* (for CSR prod at least) we're raising a ValueError (see the comment [here](https://github.com/pytorch/pytorch/blob/master/torch/_masked/__init__.py#L761-L772)). I believe the same logic applies to COO for cases with nonzero reduction identities so I followed this convention. Does that make sense?

File: torch/testing/_internal/common_methods_invocations.py (hunk 10 above)
Comment (cpuhrsch) [reply torch/testing/_internal/common_methods_invocations.py:1115]: Ah, makes sense! Thanks.

File: torch/_masked/__init__.py [hunk 11]
```diff
@@ -512,6 +512,95 @@ def _apply(a):
...
+        if reduce in {'amax', 'amin'} and new_values.size(0) == 0:
+            # IndexError: amax(): Expected reduction dim 0 to have non-zero size.
+            # sum()/prod() return the reduction identity when dim has size 0 but amax()/amin() do not
+            new_values = _reduction_identity(reduce, new_values)
+        else:
+            new_values = op(new_values, dim=0)
+        if (keepdim):
+            for _ in range(num_sparse_dims):
+                new_values = new_values.unsqueeze(0)
+        return new_values.to(dtype=output_dtype).to_sparse()
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/_masked/__init__.py:572]: You are right that torch.sparse.sum would return a dense tensor. But [`test_masked:test_masked_layout` right now expects that the layout of the output is the same as that of the input](https://github.com/pytorch/pytorch/blob/master/test/test_masked.py#L303). Additionally the [existing implementation for sparse_coo _masked.sum also tried to workaround this feature of `torch.sparse.sum`](https://github.com/pytorch/pytorch/blob/master/torch/_masked/__init__.py#L701-L708), let me know if I should change it!

[the remark this answers is not in the document]
Comment (cpuhrsch) [reply torch/_masked/__init__.py:572]: @mikaylagawarecki - ok that makes sense, thanks for digging. I guess the policy that the layouts are meant to be maintained needs to be adjusted a bit to take into account removed dimensions. We can do that as a follow-up. Can you open a small issue to track it?

File: torch/_masked/__init__.py (hunk 8 above)
Comment (cpuhrsch) [reply torch/_masked/__init__.py:786]: Just to tie things back together, I'm implementing to.dtype https://github.com/pytorch/pytorch/pull/76400

**Review by cpuhrsch — APPROVED** [review]

File: torch/_masked/__init__.py (hunk 1 above)
Comment (pearu) [reply torch/_masked/__init__.py:689]: torch.sparse.sum handles int to int64 promotion, hence, we did not need explicit promotion here.
